### DISCUSSION TITLE
Difference between Drasil and other code generators

### DISCUSSION THREAD

**Discussion by cd155** [JacquesCarette/Drasil#3070] [Ideas]
I asked this question in today's meeting, and I think it is worthwhile to have this information on the record.

While I was doing peer review, I was kind of struggling to find the difference between Drasil and other code generators. I think my problem is that I only know Drasil and barely know other code generators. Although this information will not be included in my report, it could be beneficial for many other people. 

Here are a few questions: 
- what is the selling point of Drasil? In terms of marketing Drasil, it would be great to find its uniqueness of Drasil. What makes Drasil distinguish from other code generators? 
- why Drasil is better than other code generators? What are the advantages and disadvantages?

I think this analysis is better to find the exact comparison (an example code generator). Any suggestions for selecting code generators? @smiths @JacquesCarette

**Comment by smiths** [comment]
@cd155, the biggest difference between Drasil and other generators, is that we generate more than code.  For instance, we generate documentation, like requirements specifications, and build scripts, like Makefiles.  We observed that different views of the same "knowledge" appears in multiple artifacts, so we aim to capture that information once and then transform it for every different purpose where it comes up.

For code generation itself, compared to MDA (discussed in the article by [Laurence Tratt](https://tratt.net/laurie/blog/2022/uml_my_part_in_its_downfall.html)), we generate the interface as well as the behaviour.  MDA only generates the interface from the UML diagrams.

I'm sure that when he has time @JacquesCarette can add observations about other code generators.

**Reply by cd155** [reply]
Thanks for the explanation. I wasn't fully aware of other code generators only generating code

**Reply by cd155** [reply]
My following-up question is that is there a code generator also generates code and other artifacts?

**Reply by JacquesCarette** [reply]
Some code generators do generate other code-like artifacts, but I am not aware of any that generate documentation.

**Comment by smiths** [comment]
I'm sure there are generators that build things like Makefile and test cases.  However, I don't think there is another generator that generates requirements specifications.  There might be some that call what they generate requirements, but my guess is that we are more abstract than they are with what we consider requirements.  I'd have to do more of a literature search to be more specific in my answer.  Besides the expertise of @JacquesCarette, we could talk to Dr. Mosser and Dr. Paige.

**Reply by JacquesCarette** [reply]
We really need a proper literature search. That would take quite a lot of time, unfortunately. I know I've left links to all sorts of stuff in various places on the wiki. But it's definitely not all in one place.
